specs/testing/testing.md

```diff
@@ -0,0 +1,103 @@
+
+**Note:** The testing DCRs use the format "As a ***role*** I would like to ***requirement*** so I can ***benefit***".
+
+# Roadmap
+
+Our approach is to enhance our samples, documentation and tools to provide clear guidance on how to test bots. We will also extend the SDK with new features where needed to make things easier.
+
+You can see our roadmap and work in progress stories in the [Bot Testing](https://github.com/Microsoft/BotBuilder/projects/11?fullscreen=true) project.
+
+The **4.5 Candidates** column enumerates a prioritized list of the scenarios targeted for the 4.5 release.
+
+Note that anything on here is subject to change, and is not a guarantee of shipping.
+
+# Test types
+
+There are several test types involved in bot development and operations and different authors and communities use different terminology and naming to refer to them.
+
+This section describes the types we use for testing bots. It doesn't try to provide a formal definition for each type, it is intended to serve as a broad categorization that we use to classify the different scenarios that we are supporting.
+
+- **Unit Tests**
+  
+  Are written by developers and normally executed as part of the Continuous Integration build pipeline.
+  
+  Their main purpose is to ensure that the coded logic for a bot executes as expected.
+
+- **Natural Language Understanding Tests**
+
+    Can be written by developers, NLU engineers or Product owners and can be executed as part of the CI pipeline or when the language model for the bot changes.
+
+    Their main purpose is to ensure that the bot understands what the user is asking and that there are no regressions in the language models when they are extended or modified.
+
+    These tests typically target LUIS and QnAMaker.
+
+- **Language Generation tests**
+
+   They are written by NLU engineers or Product owners and are executed as part of the CI pipeline of the underlying Language Generation files (.lg files) for the bot change.
+
+   Their main goal is to ensure the bot constructs meaningful, variable and grammatically correct responses to the user.
+
+   See [this page](https://github.com/Microsoft/botbuilder-dotnet/tree/ComposableDialog/doc/LanguageGeneration) for an overview of LG files.
+
+- **Functional tests**
+
+    Also called End to End tests, these tests target the entire bot and its dependent services.
+
+    Non-technical audiences should be able to write and execute these type of tests.
+
+- **UI Testing**
+
+    Coded UI tests are used for UI-driven functional automation of bots deployed to different channels.
+
+    These tests use open source frameworks like [Selenium](https://docs.seleniumhq.org/) or [Appium](http://appium.io/) to automate UI execution of the client app and playback pre-created scripts.
+
+    These tests are written by developers in conjuntion with product owners.
+
+- **Load Tests**
```

Comment:
We recommend bypassing the Bot Connector services for load testing: https://blog.botframework.com/2017/06/19/load-testing-a-bot/  otherwise, rate limiting is bound to be the bottleneck: https://docs.microsoft.com/en-us/azure/bot-service/bot-service-resources-bot-framework-faq?view=azure-bot-service-4.0#rate-limiting

Reply:
@damadei put together a great JMeter bot stress test package using JMeter that adheres to our guidelines: https://github.com/damadei/BotServiceStressToolkit/ instructions here: https://medium.com/microsoftazure/stress-testing-bots-in-azure-bot-service-using-apache-jmeter-part-1-3b440737fd8a

Reply:
Thanks for the feedback @EricDahlvang. I updated [this](https://github.com/Microsoft/BotBuilder/issues/5367) with your links to review them once we move forward with it.